### ISSUE TITLE
Revisit the use of this initializer

### ISSUE THREAD

**Issue by dabrahams** [val-lang/val#288]
See [this example](https://github.com/val-lang/val/pull/271/files/228d928347d87d2e544a26b2d6582483db30a931..c5adc74354f8fcc9bca928d2adc8a0df135cc9f8#diff-fb82a2ae0690aa14551a1d2dd0bf9958c20e0a0420d64614dfd94c14f1307f8cR73) and the subsequent argument.

The question is, should we make this replacement:

```diff
-    program.ast[NodeID(id)!]
+    program.ast[NodeID<ID.Subject>(rawValue: id.rawValue)] 
```

**Comment by kyouko-taiga** [comment]
What do you suggest we do?

I would be highly dissatisfied by a solution that just consist of removing the initializer that optimally converts a type-erased ID to a typed one. That's not only because of optimization, but also because it would be sad if we didn't have a way to provide an API that doesn't require a type check to be done twice every time.

Specifically, consider the following, very frequent pattern:

```swift
func foo(_ d: AnyDeclID) {
  switch d.kind {
  case FunctionDecl.self:
    foo(functionDecl: NodeID(rawVaue: d.rawValue))
  case MethodDecl.self:
    foo(methodDecl: NodeID(rawVaue: d.rawValue))
  default:
     unreachable()
  }
}

func foo(functionDecl d: NodeID<FunctionDecl>) {...}
func foo(methodDecl d: NodeID<MethodDecl>) {...}
```

In this example, I wish we would have an initializer that unconditionally converts `d` to an instance of `NodeID<FunctionDecl>` after we've already checked that `d.kind`  is indeed `FunctionDecl.self` so as to mimic what happens when we use Swift's pattern matching.

```swift
func bar(_ x: Any) {
  switch x {
    case let y as Int:
      print(y + y) // No need to re-cast `y as! Int` here.
  ...
  }
}
```
In other words, I would be disappointed if the solution was to write `NodeID(d)!`.

`NodeID.init(rawValue:)` came to fit this role by accident. It used to be an internal initializer used by the AST when we had only the raw value of an index. That use case doesn't exist anymore, so I bet one solution would be to replace `init(rawValue:)` by a new initializer `init(strawmanLabel:)` that accepts any kind of ID:

```swift
extension NodeID {
  init<T: NodeIDProtocol>(strawmanLabel other: T) {
    precondition(other.kind == Subject.kind) // or perhaps `assert`
    self.rawValue = other.rawValue
  }
}
```

**Comment by dabrahams** [comment]
> I would be highly dissatisfied by a solution that would just remove the initializer that optimally converts a type-erased ID to a typed one. That's not only because of optimization, but also because it would be sad if we didn't have a way to provide an API that doesn't require a type check to be done twice every time.

I totally understand that.  I don't think the un-checked constructor should be unavailable: it's a basic principle of generic programming to expose an _efficient_ set of “basis operations” available for every type.  I just think this one should only be used where profiling tells us it's needed for optimization purposes.  We are primarily building an application here; it's not a goal (yet) to produce an optimal library of generic algorithms. I don't know where "twice" comes from in this case, but I'd lay a strong wager that the optimizer will eliminate redundant checks in common situations.

I think your usage pattern problem is real, but I don't think that initializer is a great way to address it.  In fact I think we're talking about the same thing in #289, aren't we?

**Comment by kyouko-taiga** [comment]
Yes, #289 is strongly related. The "twice" comes from the fact that:
1. we need to verify that `x` has kind `K` to enter some case;
2. once inside the case, we need to "cast" `x` as `NodeID<K>`, causing a second test if we use `NodeID.init(_:)`.

**Comment by dabrahams** [comment]
That sounds like exactly the kind of check that will be eliminated by the optimizer.

**Comment by dabrahams** [comment]
It can, because we are not building with resilience enabled.  We can of course go further and mark the initializer `@inlinable` or `@inline(__always)`
